op-program/client/l2/oracle.go

```diff
@@ -26,3 +31,41 @@ type Oracle interface {
 	// BlockByHash retrieves the block with the given hash.
 	BlockByHash(blockHash common.Hash) *types.Block
 }
+
+// PreimageOracle implements Oracle using by interfacing with the pure preimage.Oracle
+// to fetch pre-images to decode into the requested data.
+type PreimageOracle struct {
+	oracle preimage.Oracle
+	hint   preimage.Hinter
+}
+
+var _ Oracle = (*PreimageOracle)(nil)
+
+func NewPreimageOracle(raw preimage.Oracle, hint preimage.Hinter) *PreimageOracle {
+	return &PreimageOracle{
+		oracle: raw,
+		hint:   hint,
+	}
+}
+
+func (p *PreimageOracle) BlockByHash(blockHash common.Hash) *types.Block {
+	p.hint.Hint(L2BlockHint(blockHash))
+	blockRlp := p.oracle.Get(preimage.Keccak256Key(blockHash))
```

Comment:
One important thing to note here - `blockHash` is actually the hash of the block _header_ not the full block (as is always the case in Ethereum). So we'll have to retrieve the block header RLP then get the transactions based on the TxRoot and then reassemble the block. We can hard code empty uncles and withdrawals.